### PR TITLE
feat(facial-expressions): add the facial expression feature and display them in speakerstats

### PR DESCRIPTION
**Pull request by gabiborlea**
Adds the facial expression feature. It uses the face-api.js library which is used in a web worker. The web worker receives image data that comes from a canvas created from the local video track. The web worker sends back the expression and then the main thread will send the expression to the other participants via data channel and to the xmpp sever, also saving it to the redux state. Then the expressions from all participants will be shown in the speaker stats dialog.  

Demo at: https://gborlea.jitsi.net/
Screenshot 2021-11-02 at 17 31 16

### CODE REVIEW

**Comment by damencho** [comment]
Nice one!

How often is this beeing sent to the backend?

**Comment by damencho** [comment]
Maybe we will need a config.js property to disable this functionality

**Comment by damencho** [comment]
Jenkins add to whitelist.

**Comment by gabiborlea** [comment]
The detection is done every 1000ms when using webgl and every 3000ms when using cpu, and the last facial expression with its duration is sent to the backend when a new one that is not the same with the last one occurs. 
And if the disableFacialRecognition flag is set to true in the config.js it will be disabled.

**Comment by damencho** [comment]
> The detection is done every 1000ms when using webgl and every 3000ms when using cpu, and the last facial expression with its duration is sent to the backend when a new one that is not the same with the last one occurs.

In your tests how often the change is triggered?

Hum, but wait how is my expression propagated to the other participants?

And about the config, you haven't documented it in config.js or I missed that, check how we have commented the default value with some simple explanation what it does.

**Comment by gabiborlea** [comment]
I managed to change the facial expression every second for like 4 or 5 seconds, and the expression was sent to the backend every second, but this would be the worst-case as it is not natural to change your facial expression every second. 
The expression is propagated to the other participants during the conference via the data channel. When the user joins the conference he will receive the facial expressions of the other participant that occurred prior to his join from the prosody.
Sorry, I did not document the flag in config.js.

**Comment by gabiborlea** [comment]
I forgot to mention that you can test the feature at https://gborlea.jitsi.net/.
If you open the speaker stats dialog it should look like this.
Screenshot 2021-09-27 at 10 10 53

**Review by quitrk — CHANGES_REQUESTED** [review]

File: react/features/facial-recognition/middleware.js [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+    case CONFERENCE_JOINED: {
+        const { getState, dispatch } = store;
+        const { enableFacialRecognition } = getState()['features/base/config'];
+
+        if (!enableFacialRecognition) {
+            return next(action);
+        }
+
+        dispatch(loadWorker());
+        dispatch(setFacialRecognitionAllowed(true));
```
[the remark this answers is not in the document]
Comment (gabiborlea) [reply react/features/facial-recognition/middleware.js:32]: I used that flag for preventing the recognition from starting in the prejoin screen

File: react/features/facial-recognition/actions.js [hunk 2]
```diff
@@ -0,0 +1,261 @@
...
+                } else {
+                    lastFacialExpression
+                    && dispatch(addFacialExpression(lastFacialExpression, duplicateConsecutiveExpressions + 1));
+                    lastFacialExpression = value;
+                    duplicateConsecutiveExpressions = 0;
+                }
+            }
+            const { detectionTimeInterval } = state['features/facial-recognition'];
+
+            sendDataToWorker(worker, imageCapture, detectionTimeInterval);
```
[the remark this answers is not in the document]
Comment (gabiborlea) [reply react/features/facial-recognition/actions.js]: Fixed it.

File: react/features/facial-recognition/reducer.js [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+        };
+    }
+    case SET_DETECTION_TIME_INTERVAL: {
+        return {
+            ...state,
+            detectionTimeInterval: action.time
+        };
+    }
+    case UPDATE_CAMERA_TIME_TRACKER: {
+        if (action.muted && state.cameraTimeTracker.lastCameraUpdate !== 0) {
```
[the remark this answers is not in the document]
Comment (gabiborlea) [reply react/features/facial-recognition/reducer.js]: Removed it.

File: react/features/facial-recognition/middleware.js [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+import { sendFacialExpressionToParticipants, sendFacialExpressionToServer } from './functions';
+
+
+MiddlewareRegistry.register(store => next => action => {
+    switch (action.type) {
+    case CONFERENCE_JOINED: {
+        const { getState, dispatch } = store;
+        const { enableFacialRecognition } = getState()['features/base/config'];
+
+        if (!enableFacialRecognition) {
```
[the remark this answers is not in the document]
Comment (gabiborlea) [reply react/features/facial-recognition/middleware.js]: Fixed it.

**Review by quitrk — DISMISSED** [review]

**Comment by quitrk** [comment]
jenkins please test this please

**Comment by gabiborlea** [comment]
jenkins please test this please

**Comment by saghul** [comment]
Jenkins add to whitelist.

**Comment by saghul** [comment]
Jenkins please test this please.

**Comment by gabiborlea** [comment]
Jenkins please test this please.

**Review by quitrk — APPROVED** [review]

**Comment by gabiborlea** [comment]
Jenkins please test this please.

**Review by andrei-gavrilescu — APPROVED** [review]

**Pull request merged**